url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/layout/hierarchical/model/mxGraphHierarchyNode.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2005, David Benson
+ *
+ * All rights reserved.
+ *
+ * This file is licensed under the JGraph software license, a copy of which
+ * will have been provided to you in the file LICENSE at the root of your
+ * installation directory. If you are unable to locate this file please
+ * contact JGraph sales for another copy.
+ */
+package com.mxgraph.layout.hierarchical.model;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * An abstraction of an internal node in the hierarchy layout
+ */
+public class mxGraphHierarchyNode extends mxGraphAbstractHierarchyCell
+{
+
+	/**
+	 * Shared empty connection map to return instead of null in applyMap.
+	 */
+	public static Collection<mxGraphHierarchyEdge> emptyConnectionMap = new ArrayList<mxGraphHierarchyEdge>(0);
+
+	/**
+	 * The graph cell this object represents.
+	 */
+	public Object cell = null;
+
+	/**
+	 * Collection of hierarchy edges that have this node as a target
+	 */
+	public Collection<mxGraphHierarchyEdge> connectsAsTarget = emptyConnectionMap;
+
+	/**
+	 * Collection of hierarchy edges that have this node as a source
+	 */
+	public Collection<mxGraphHierarchyEdge> connectsAsSource = emptyConnectionMap;
+
+	/**
+	 * Assigns a unique hashcode for each node. Used by the model dfs instead
+	 * of copying HashSets
+	 */
+	public int[] hashCode;
+
+	/**
+	 * Constructs an internal node to represent the specified real graph cell
+	 * @param cell the real graph cell this node represents
+	 */
+	public mxGraphHierarchyNode(Object cell)
+	{
+		this.cell = cell;
+	}
+
+	/**
+	 * Returns the integer value of the layer that this node resides in
+	 * @return the integer value of the layer that this node resides in
+	 */
+	public int getRankValue()
+	{
+		return maxRank;
+	}
+
+	/**
+	 * Returns the cells this cell connects to on the next layer up
+	 * @param layer the layer this cell is on
+	 * @return the cells this cell connects to on the next layer up
+	 */
+	@SuppressWarnings("unchecked")
+	public List<mxGraphAbstractHierarchyCell> getNextLayerConnectedCells(int layer)
+	{
+		if (nextLayerConnectedCells == null)
+		{
+			nextLayerConnectedCells = new ArrayList[1];
+			nextLayerConnectedCells[0] = new ArrayList<mxGraphAbstractHierarchyCell>(connectsAsTarget.size());
+			Iterator<mxGraphHierarchyEdge> iter = connectsAsTarget.iterator();
+			
+			while (iter.hasNext())
+			{
+				mxGraphHierarchyEdge edge = iter.next();
+
+				if (edge.maxRank == -1 || edge.maxRank == layer + 1)
+				{
+					// Either edge is not in any rank or
+					// no dummy nodes in edge, add node of other side of edge
+					nextLayerConnectedCells[0].add(edge.source);
+				}
+				else
+				{
+					// Edge spans at least two layers, add edge
+					nextLayerConnectedCells[0].add(edge);
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 82:97, 118:132.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.